libs/blocks/library-config/lists/blocks.js

```diff
@@ -51,23 +57,154 @@ function getTable(block, name, path) {
+  if (!container || !container.elements) return '';
+  const firstBlock = container.elements[0];
+  if (container[LIBRARY_METADATA]) {
+    const libraryMetadata = getMetadata(container[LIBRARY_METADATA]);
+    return libraryMetadata?.searchtags?.text
+      ? `${libraryMetadata?.searchtags?.text} ${getBlockName(firstBlock)}`
+      : getBlockName(firstBlock);
   }
-  const libraryMetadata = getMetadata(block.nextElementSibling);
-  return libraryMetadata?.searchtags?.text
-    ? `${libraryMetadata?.searchtags?.text} ${getBlockName(block)}`
-    : getBlockName(block);
+  return getBlockName(firstBlock);
 }
 
-function isMatchingBlock(pageBlock, query) {
-  const tagsString = getBlockTags(pageBlock);
+export function isMatching(container, query) {
+  const tagsString = getSearchTags(container);
   if (!query || !tagsString) return false;
   const searchTokens = query.split(' ');
   return searchTokens.every((token) => tagsString.toLowerCase().includes(token.toLowerCase()));
 }
 
+export function getContainers(doc) {
+  /* A page describing blocks is assumed to have the following representation
+  (e.g. the page at /docs/library/blocks/carousel.plain.html):
+  <body>
+    <div>
+      <p>Single block container</p>
+      <h2>...</h2>
+      <div class="block1">
+      <div class="block2">
+      <div class="library-metadata">
+    </div>
+    <div>
+      <h2>...</h2>
+      <p>Multiple block container</p>
+      <div class="library-container-start"></div>
+      <p>...</p>
+      <div class="block3"></div>
+    </div>
+    ...
+    <div>
+      <h2>...</h2>
+      <p>...</p>
+      <div class="block4"></div>
+      <div class="library-container-end"></div>
+      <div class="library-metadata">
+      ...
+    </div>
+    ...
+  </body>
+
+  The page html is parsed into sections (div children of body) and
+  sub-sections (children of section).
+  Parsing the above html results in the following container array:
+  [{
+      elements: [<div class="block1">],
+    }, {
+      elements: [<div class="block2">],
+      library-metadata: <div class="library-metadata">
+    }, {
+      elements: [<p>, <div class="block3">, <p>---</p>, ...,<h2>, <p>, <div class="block4">],
+      library-metadata: <div class="library-metadata">
+    },
+    ...
+  ]
+   */
+  if (!doc || !doc.body) return [];
+  const sections = doc.body.children;
+  if (sections.length === 0) return [];
+  const containers = [];
+  const sectionBreak = doc.createElement('p');
+  sectionBreak.innerHTML = '---';
+  let container = { elements: [] };
+  let withinContainer = false;
+  for (let i = 0; i < sections.length; i += 1) {
+    const section = sections[i];
+    const subSections = section.children;
+    // eslint-disable-next-line no-continue
+    if (subSections.length === 0) continue;
+    for (let j = 0; j < subSections.length; j += 1) {
+      const subSection = subSections[j];
+      if (subSection.className === LIBRARY_CONTAINER_START) {
```

Comment:
This `if/else` block looks quite complex and hard to read. Any chance of refactoring it to use `switch/else` statement for example?

Reply:
Addressed in https://github.com/adobecom/milo/pull/772/commits/e59df2f53ee5fade316b26cabb26adfb6191362a